pkg/schema/tracker.go

```diff
@@ -329,3 +340,208 @@ func (tr *Tracker) CreateTableIfNotExists(table *filter.Table, ti *model.TableIn
+			return nil, err
+		}
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		tr.downstreamTrack[tableID] = createTableStmt
+	}
+
+	// get PK/UK from  Constraints.
+	var index *model.IndexInfo
+	for _, constraint := range createTableStmt.Constraints {
+		var keys []*ast.IndexPartSpecification
+
+		switch constraint.Tp {
+		case ast.ConstraintPrimaryKey: // pk.
+			keys = constraint.Keys
+		case ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex: // unique,unique key,unique index.
+
+			if index == nil {
+				keys = constraint.Keys
+			} else {
+				// if index has been found, uk should be jump.
+				continue
+			}
+		default:
+			continue
+		}
+
+		if keys != nil {
+			columns := make([]*model.IndexColumn, 0, len(keys))
+			isAllNotNull := true // pk is true, uk should check.
+			for _, key := range keys {
+
+				// UK should check not null.
+				if constraint.Tp != ast.ConstraintPrimaryKey {
+
+					for _, column := range createTableStmt.Cols {
+						if key.Column.Name.String() == column.Name.String() {
+							hasNotNull := false
+							for _, option := range column.Options {
+								if option.Tp == ast.ColumnOptionNotNull {
+									hasNotNull = true
+									break
+								}
+							}
+							if !hasNotNull {
+								isAllNotNull = false
+							}
+							break
+						}
+					}
+				}
+
+				if !isAllNotNull {
+					break
+				}
+
+				if orginColumn := model.FindColumnInfo(originTi.Columns, key.Column.Name.O); orginColumn != nil {
+					column := &model.IndexColumn{
+						Name:   key.Column.Name,
+						Offset: orginColumn.Offset,
+						Length: key.Length,
+					}
+					columns = append(columns, column)
+				}
+			}
+
+			if !isAllNotNull {
+				continue
+			}
+
+			if len(columns) != 0 {
+				if constraint.Tp == ast.ConstraintPrimaryKey {
+					index = &model.IndexInfo{
+						Table:   createTableStmt.Table.Name,
+						Unique:  true,
+						Primary: true,
+						State:   model.StatePublic,
+						Tp:      model.IndexTypeBtree,
+						Columns: columns,
+					}
+					log.L().Debug(fmt.Sprintf("Find DownStream table %s pk %s", tableID, constraint.Name))
+					return index, nil // pk > uk.
+				}
+				// uk should continiue to find pk.
+				index = &model.IndexInfo{
+					Table:   createTableStmt.Table.Name,
+					Unique:  true,
+					Primary: false,
+					State:   model.StatePublic,
+					Tp:      model.IndexTypeBtree,
+					Columns: columns,
+				}
+				log.L().Debug(fmt.Sprintf("Find DownStream table %s uk(not null) %s", tableID, constraint.Name))
+			}
+		}
+	}
+
+	if index == nil {
+		log.L().Debug(fmt.Sprintf("DownStream table %s has no pk/uk(not null)!", tableID))
+	}
+
+	return index, nil
+}
+
+// GetAvailableDownStreanUKIndexInfo gets available downstream UK whose data is not null.
+// note. this function will not init downstreamTrack.
+func (tr *Tracker) GetAvailableDownStreanUKIndexInfo(tableID string, originTi *model.TableInfo, data []interface{}) *model.IndexInfo {
```

Comment:
What's the difference between this function and 	`get PK/UK from  Constraints` in the front?

Reply:
The purpose of this function is to find an uk which data is not null at current row.
It will be called when `get PK/UK from Constraints` can not find an UK which has `not null` Constraints.